### PR TITLE
Use computed visibility in rustdoc

### PR DESCRIPTION
**Pull request by inquisitivecrystal**
This PR changes `librustdoc` to use computed visibility instead of syntactic visibility. It was initially part of #88019, but was separated due to concerns that it might cause a regression somewhere we couldn't predict.

r? @jyn514
cc @cjgillot @petrochenkov

### CODE REVIEW

**Comment by jyn514** [comment]
> There are a few other places in librustdoc that should be checked to see if they can use computed visibility. However, I don't understand them, so they're best left to someone who knows what they're doing.

Can you open an issue for this so it doesn't get forgotten? Ideally with links to the relevant code :) but no worries if that turns out to be a pain.

**Comment by jyn514** [comment]
This should fix the issue @petrochenkov mentioned in https://github.com/rust-lang/rust/pull/88019#issuecomment-898962228, right, macro_export macros should now be displayed with the proper visibility? Can you add a test for that?

**Comment by inquisitivecrystal** [comment]
> > There are a few other places in librustdoc that should be checked to see if they can use computed visibility. However, I don't understand them, so they're best left to someone who knows what they're doing.
> 
> Can you open an issue for this so it doesn't get forgotten? Ideally with links to the relevant code :) but no worries if that turns out to be a pain.

It turns out I spoke too soon! There are fewer instances than I thought, and I should be able to fix all of them. The reason I thought they were above my paygrade was that I tried fixing one and it made tests fail, but it looks like I was just doing something wrong? It would still be good to have you check them over, but I presume you'd do that anyway.

> This should fix the issue @petrochenkov mentioned in #88019 (comment), right, macro_export macros should now be displayed with the proper visibility? Can you add a test for that?

That already works, since I did a workaround in that PR (specifically, I checked for the macro_export directly). Do you still want me to do a test?

Oh, also, I think we need an `is_public` method on `ty::Visibility`. It seems to pop up all over the place. Should I add that here, or should that be a seperate PR?

**Comment by jyn514** [comment]
> Do you still want me to do a test?

Yes, please.

> Oh, also, I think we need an is_public method on ty::Visibility. It seems to pop up all over the place. Should I add that here, or should that be a seperate PR?

Here is fine.

**Comment by jyn514** [comment]
@inquisitivecrystal do you know when you'll get a chance to look at this?

**Comment by inquisitivecrystal** [comment]
> @inquisitivecrystal do you know when you'll get a chance to look at this?

Sorry for dropping off the face of the earth! I've been dealing with a lot of stuff IRL. 

I don't know exactly when I'll have a chance to work on this again, but I'm hoping it'll be in the next few weeks. The sooner I can find the time the happier I'll be.

**Comment by inquisitivecrystal** [comment]
Okay, first off, I apologize for taking such a long amount of time to get back to this.

> > Do you still want me to do a test?
> 
> Yes, please.

I've gotten myself kind of confused. (I'll also mention that most of this is drawn from notes I wrote up months ago, which probably isn't helping.)

The relevant line, as adjusted by #88019, is here:

https://github.com/rust-lang/rust/blob/c1026539bd22e9d070988deaa47b1360cbc76436/src/librustdoc/visit_ast.rs#L257

This checks whether an AST item is public by checking its syntactic visibility. After this PR, it would instead use computed visibility. `is_pub` is then used in a few places:

https://github.com/rust-lang/rust/blob/c1026539bd22e9d070988deaa47b1360cbc76436/src/librustdoc/visit_ast.rs#L259-L261
https://github.com/rust-lang/rust/blob/c1026539bd22e9d070988deaa47b1360cbc76436/src/librustdoc/visit_ast.rs#L271
https://github.com/rust-lang/rust/blob/c1026539bd22e9d070988deaa47b1360cbc76436/src/librustdoc/visit_ast.rs#L287-L293

The first one seems irrelevant to macros. All `store_path` does is write to `exact_paths`, which is [only used for creating implementors sections](https://github.com/rust-lang/rust/blob/master/src/librustdoc/formats/cache.rs#L48-L58). Implementors sections have nothing to do with macros. The third one also seems irrelevant, since it only concerns `use` declarations, though I'm less sure in that case. 

The second case is applicable to macros. It looks like what it's doing is making sure that only public items are ever inlined, which makes sense, since they're the only items that can be imported via a use statement. This is backed up by the test [`inline_local/macro_by_example.rs`](https://github.com/rust-lang/rust/blob/c1026539bd22e9d070988deaa47b1360cbc76436/src/test/rustdoc/inline_local/macro_by_example.rs). This effectively tests whether `#[macro_export]` macros are treated as public for inlining (they are). The test fails if I remove the `#[macro_export] special casing from `is_pub`, though it starts working again once I switch to computed visibility.

I think this test may cover everything, but I'm not certain. Am I missing or misunderstanding anything?

**Review by jyn514 — COMMENTED** [review]
Your writeup is convincing to me, I think the existing test is fine :) just one question and then I think we can merge this.

File: src/librustdoc/visit_ast.rs [hunk 1]
```diff
@@ -103,15 +99,15 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
 
     fn visit_mod_contents(
         &mut self,
-        vis: &hir::Visibility<'_>,
         id: hir::HirId,
         m: &'tcx hir::Mod<'tcx>,
         name: Symbol,
     ) -> Module<'tcx> {
         let mut om = Module::new(name, id, m.inner);
+        let def_id = id.expect_owner();
```
[the remark this answers is not in the document]
Comment (inquisitivecrystal) [reply src/librustdoc/visit_ast.rs:107]: I'm afraid I'm not really familiar with the method you're referring to? I couldn't find anything like that, but that doesn't mean much, given how much of a maze `TyCtxt` is. In any case, wouldn't it be better to check the invariant that the module is its own owner rather than assuming it?

[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/visit_ast.rs:107]: Ah, it's `tcx.hir().local_def_id_to_hir_id()`.

> In any case, wouldn't it be better to check the invariant that the module is its own owner rather than assuming it?

If modules aren't their own owners, then basically all of rustdoc's code is broken and every test will fail. I don't think we need to assert it for every module rustdoc ever encounters.

**Comment by inquisitivecrystal** [comment]
For the last few days I've been having a look at expanding this to all of rustdoc, as we discussed above a while ago. I'm sorry I didn't update you on my status. I think I've almost got it working, but there's this one really tricky problem I don't know how to deal with. 

https://github.com/rust-lang/rust/blob/5fa94f3c57e27a339bc73336cd260cd875026bd1/src/librustdoc/clean/mod.rs#L1995-L2007

On that second line, it checks whether the syntactic visibility is either public or restricted. In the HIR, [there are four `VisibilityKinds`: `Public`, `Crate`, `Restricted`, and `Inherited`](https://github.com/rust-lang/rust/blob/c4ff03f689c03d73d3a8af49611b1f2b54d6a300/compiler/rustc_hir/src/hir.rs#L2561-L2566). At the level of computed visibility, [there are only three visibilities: `Public`, `Restricted`, and `Invisible`](https://github.com/rust-lang/rust/blob/a49e38e672c60da788360e088f00ad12353e3913/compiler/rustc_middle/src/ty/mod.rs#L175-L182). My first attempt was to do a direct translation, where a `VisibilityKind` of public or restricted would be equivalent to a computed visibility of public or restricted. However, that's not accurate.

https://github.com/rust-lang/rust/blob/a49e38e672c60da788360e088f00ad12353e3913/src/test/rustdoc/reexports-priv.rs#L50-L53

In the case of this test, the use has a `VisibilityKind` of `Inherited`, but a computed visibility of `Restricted`. Note that the conversion is lossy: in this case, `pub(self) use` would have had a `VisibilityKind` of `Restricted` but the same computed visibility. Thus, in document private items mode, rustdoc currently documents the `pub(self) use` version but not the version here, despite them being equivalent in meaning. At least, if I'm understanding the situation correctly, which I very well might not be.

How should I handle this?

File: src/librustdoc/visit_ast.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (inquisitivecrystal) [reply src/librustdoc/visit_ast.rs:107]: Alright!

**Comment by jyn514** [comment]
Ugh, that logic is painful. It's trying to predict ahead of time whether the item will be stripped, but before the tcx is available and using the AST visibility.

> Thus, in document private items mode, rustdoc currently documents the pub(self) use version but not the version here, despite them being equivalent in meaning. 

That definitely seems like a bug, it should be consistent between the two. @GuillaumeGomez you added this in https://github.com/rust-lang/rust/pull/86841/ - what was your goal in sometimes not inlining here?

**Comment by jyn514** [comment]
@inquisitivecrystal I think it makes sense to land your other changes in the meantime if you want to push them :)

**Comment by GuillaumeGomez** [comment]
#86841 was because we have different macro kinds with different visibility settings. The macro 2.0 can be handled like any items whereas the old macros can't. And because of that, we have this weird handling. If it can be improved, I'm all for it! :D

**Comment by inquisitivecrystal** [comment]
> #86841 was because we have different macro kinds with different visibility settings. The macro 2.0 can be handled like any items whereas the old macros can't. And because of that, we have this weird handling. If it can be improved, I'm all for it! :D

I think #88019 means they can be treated the same now. Does that open up any directions for improvement?

**Comment by jyn514** [comment]
@GuillaumeGomez I don't really understand your response; what was the goal of adding the special casing? What does it have to do with macros?

**Comment by GuillaumeGomez** [comment]
The issue fixed in #86841 was for the macro 2.0 and a small cleanup. Unless I missed something? Sorry if my explanations aren't clear... Let me try again:

We weren't rendering macros 2.0 correctly because we handled them like the old macros (their visibility too). So I simply changed how we handled them so it was working as expected.

There wasn't big changes in how we handle visibility as far as I can see but maybe I missed something?

**Comment by jyn514** [comment]
@GuillaumeGomez I'm asking specifically about this change: https://github.com/rust-lang/rust/pull/86841/files#diff-384affc1b4190940f114f3fcebbf969e7e18657a71ef9001da6b223a036687d9R2057
It seems unrelated to either macros or the rest of the PR.

**Comment by GuillaumeGomez** [comment]
Oh I see. Sorry, missed it. So for this one, the goal was to fix the display of reexported items if the "--document-private-items" option is used (it's described as is in the PR description). However in this case it seems like it's impacting all items, not just the reexported ones.

**Comment by jyn514** [comment]
> However in this case it seems like it's impacting all items, not just the reexported ones.

I don't understand the distinction you're making between "all items" (when this is a bit of code only dealing with `use` items) and "re-exported items". What was the behavior before your PR? Was `pub(self)` still shown? I'm a little dubious how useful inlining that is.

**Comment by GuillaumeGomez** [comment]
I think it was but not for all items (still based on the PR description). Typedef and reexports didn't as far as what is described. Sorry, that's all I remember...

**Comment by jyn514** [comment]
Ok. Given that you don't remember why you made the change and it's very hard to preserve the behavior, @inquisitivecrystal I think it's fine to just remove that check.

**Comment by GuillaumeGomez** [comment]
Checking what tests it breaks will be simpler indeed.

**Comment by GuillaumeGomez** [comment]
Absolutely: `pub(self)` doesn't bring any useful information. I think this change was mostly for `pub(crate)` or equivalent (it makes mostly sense to display them when documenting private items). Anyway, I guess we'll see soon enough. :)

**Comment by inquisitivecrystal** [comment]
> Ok. Given that you don't remember why you made the change and it's very hard to preserve the behavior, @inquisitivecrystal I think it's fine to change the check to not inline either bare `use` items or `pub(self) use` items. So the check would be something like `vis.is_pub() || matches!(vis, Visibility::Restricted(mod_) if mod_ != current_mod)`.

Well, strictly speaking, it isn't particularly hard to preserve the existing behavior. The problems with the current behavior are twofold: it is very confusing and it relies on syntactic visibility. 

How is this: `visibility.is_at_least(ty::Visibility::Restricted(parent_module.to_def_id()), cx.tcx)))`? I think that's the non-hacky way of doing things, if we've established that it's fine to treat `pub(self)` as equivalent to no modifier.

**Comment by jyn514** [comment]
@inquisitivecrystal that sounds great! :)

**Comment by inquisitivecrystal** [comment]
> @inquisitivecrystal that sounds great! :)

I got it down further! `visibility.is_accessible_from(parent_mod.to_def_id(), cx.tcx)` does the same thing, but is shorter and more readable. 😄

Anyway, pushing now!

**Comment by jyn514** [comment]
> visibility.is_accessible_from(parent_mod.to_def_id(), cx.tcx) does the same thing

Hmm, are you sure? That makes me worried we'll run into https://github.com/rust-lang/rust/issues/64762. Can you add a test with a nested import?

**Comment by inquisitivecrystal** [comment]
> > visibility.is_accessible_from(parent_mod.to_def_id(), cx.tcx) does the same thing
> 
> Hmm, are you sure? That makes me worried we'll run into #64762. Can you add a test with a nested import?

Am I sure that `visibility.is_accessible_from(parent_mod.to_def_id(), cx.tcx)` is equivalent to `vis.is_at_least(ty::Visibility::Restricted(parent_mod.to_def_id()), cx.tcx)`? Yes, fairly, after reading [the definition](https://github.com/rust-lang/rust/blob/8b09ba6a5d5c644fe0f1c27c7f9c80b334241707/compiler/rustc_middle/src/ty/mod.rs#L317-L326).

I'll confess I don't grasp how #64762 could affect this. Could you explain what that would look like and what I should be testing for? I'm sorry I keep having so many questions.

**Review by jyn514 — APPROVED** [review]
The rustdoc changes LGTM. @petrochenkov how do you feel about the switch from `vis == Visibility::Public` to `vis.is_pub()` in the resolver?

**Comment by petrochenkov** [comment]
Mostly neutral, `vis == Visibility::Public` was already short enough, but I guess it's common enough to introduce a slightly shorter method instead.
r? @jyn514

**Review by petrochenkov — APPROVED** [review]

**Comment by nnethercote** [comment]
It was suggested that this may have caused a large regression in `doc` times for the `webrender-wrench` benchmark: https://github.com/rust-lang/rust/pull/90769#issuecomment-971527515. Any thoughts about whether that's true?

**Comment by jyn514** [comment]
That is ... Quite a large difference, wow.

I wouldn't expect this to be the cause of *all* the regressions there, this shouldn't have affected get_blanket_impls or evaluate_obligation. I think it's plausible it makes clean_crate slower but that's not where most of the regressions are.

Maybe open a couple PRs reverting changes in the rollup and see which if any fix the regression?

**Comment by jyn514** [comment]
https://github.com/rust-lang/rust/pull/88868 affects codegen, rolling that up seems suspicious. That or
https://github.com/rust-lang/rust/pull/90727 would be my guess (altogether neither of those should have affected evaluate_obligation quite so much ...)

**Comment by jyn514** [comment]
https://github.com/rust-lang/rust/pull/90361 also seems plausible.

**Comment by camelid** [comment]
Followup: IIUC, the perf regression was caused by the behavior regression in #90865. Then #91094 fixed the behavior regression, which in turn fixed the perf regression.
